Perf

mredaelli opened the pull request:
A few changes

The result is going from
```
satisfia (b28010a) 📦🤷 🐍 v3.10.13 via ⊢💩 (shell) on ☁️  m.redaelli@gmail.com
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     38.970 s ±  0.577 s    [User: 38.611 s, System: 0.551 s]
  Range (min … max):   38.037 s … 39.630 s    10 runs
```
to
```
satisfia on 🌱 perf 📦🤷 🐍 v3.10.13 via ⊢💩 (shell) on ☁️  m.redaelli@gmail.com
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     30.802 s ±  1.359 s    [User: 30.457 s, System: 0.587 s]
  Range (min … max):   29.625 s … 33.938 s    10 runs
```

last commit goes down to 24 secs

mredaelli commented on src/satisfia/agents/makeMDPAgentSatisfia.py:
```diff
@@ -815,54 +817,57 @@ def randomTieBreaker(self, state, action):
 
 	# state, action, aleph4state, aleph4action, estActionProbability
-	#@lru_cache(maxsize=None)
+	@lru_cache(maxsize=None)
```
this was disabled but it seems to help.

was there a reason why you disabled it?

If it was a memory problem, it might be worth to put a limit, like 1024: it would probably still help

mredaelli commented on src/satisfia/agents/makeMDPAgentSatisfia.py:
```diff
@@ -1380,6 +1385,7 @@ def X(nextState):
 			print(pad(state),"| | | | | ╰ ETerminalState2_action", prettyState(state), action, aleph4action, policy, ":", res)
 		return res
 
+	@cache
```
this also was not cached. Any reason against it?

mensch72 replied:
i might simply have forgotten it

mredaelli commented on src/satisfia/util/distribution.py:
```diff
@@ -40,7 +38,7 @@ def var(self, *, precision=64):
 
 	def stddev(self, *, precision=64):
-		return torch.sqrt(self.var(precision))
+		return math.sqrt(self.var(precision))
```
would you mind this? we don't need it yet, and it messes up with browser support

mensch72 replied:
ok for now, at some point we might however need it for training.
what about 
```
try:
  import torch
  sqrt = torch.sqrt
except:
  sqrt = math.sqrt  # or numpy?
```

mredaelli replied:
My guess would be that torch can be relegated to files that are imported only from the training part, which in turn are not imported by the web book, so no problem arises.

If that turns out not to be true, yes, your solution should work

mredaelli commented on src/satisfia/util/helper.py:
```diff
@@ -1,56 +1,42 @@
+    else:
+        return _Interval(left,  left if right is None else right)
+
+class _Interval(NamedTuple):
```
namedtuples are much faster than generic classes.

This change should be transparent to the rest of the code

mensch72 replied:
very good!

mredaelli commented:
Caching also `propagateAspiration` brings it down to 29s, and the memory usage for stays below 450MB.

mredaelli commented:
Next things to consider would be:

- `_extract_state_attributes` takes a lot of time. If `state` instead were already a tuple, we'd save all that computation plus the time to call the function. Would that be messy?
- `transition_distribution`

mredaelli commented:
I pushed a cleanup of the utilities, with the interval version and the "scalar but maybe second argument is an interval" version:

```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     26.851 s ±  0.742 s    [User: 26.346 s, System: 0.600 s]
  Range (min … max):   26.273 s … 28.715 s    10 runs
```

so we are saving about 1/3 of the time.

I actually specialized the utilities a bit more, adding a scalar version (e.g. `clip`), a version where only the central argument is a tuple (`clip2`) and a generic one for intervals (`clipInt`), the original one, which however doesn't seem to be used.

That saves a bit more, and I like it better because it's more... "typed" :), but I understand if it gets in the way:
```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
 ⠦ Current estimate: 26.614 s     ████████████████████████████░░░░░░░░░░░░░░░░░░░░ ETA 00:01:46  Time (mean ± σ):     26.436 s ±  0.631 s    [User: 25.798 s, System: 0.607 s]
  Range (min … max):   25.509 s … 27.904 s    10 runs
```

mredaelli commented:
Latest commit shaves another couple of seconds by excluding from the loss computation stuff with 0 loss parameter directly in `__init__`. 

Losses for time and `lossCoeff4OtherLoss` still need adaptation though.

I also replaces all tabs with spaces because I'm going crazy with whitespace errors. I hope that's OK. To see the actual difference, disable whitespace in the diff config:

![image](https://github.com/pik-gane/satisfia/assets/3073833/96a9d684-e1fc-4039-b636-582eddb16b3e)

mensch72 commented on src/satisfia/agents/makeMDPAgentSatisfia.py:
```diff
@@ -17,1546 +17,1522 @@
+        if self.debug:
+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,"...")
+
+        res = sum(k * v(s, a, al4s, al4a, p) for k, v in self.losses.items())
```
shouldn't we also make sure that `v` is only called if `k != 0`, like so: `0 if k == 0 else k * v(...)`?

mredaelli replied:
If I'm not mistaken, that's done in `__init__`, here:

https://github.com/pik-gane/satisfia/pull/7/files#diff-2a77c67961414a5d12c1a7db8c3b4536e945a06643c4f16f68f3509b002378eeR190

so we do it only once rather than at every evaluation.

Unless the parameters can be changed?

mensch72 commented on src/satisfia/agents/makeMDPAgentSatisfia.py:
```diff
@@ -17,1546 +17,1522 @@
+
+        if self.verbose or self.debug:
+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,":",res,"\n"+pad(s),"| |    ")
+
```
I think we should print the nonzero loss components at least if `debug==True`

mredaelli replied:
True, that needs to be done. If you agree with the rest, I'll add it

mensch72 commented on src/satisfia/util/helper.py:
```diff
@@ -1,84 +1,81 @@
-def interpolate(x, l, y):
-	# denoted	x : l : y in formulas
+def relativePosition2(x, z, y):
+    if (diff := (y-x)) == 0.:
```
Interesting, I wasn't aware of the `:=` operator in Python. Good to know...

simfis replied:
It's pretty new: https://peps.python.org/pep-0572/

mensch72 commented on src/world_model/world_model.py:
```diff
@@ -62,27 +62,26 @@ def default_policy(self, state):
-    def possible_successors(self, state, action=None, n_samples = None):
-        """Return a list of possible successor states after performing action in state,
+    def possible_successors(self, state, action=None, n_samples = None) -> set:
+        """Return the of possible successor states after performing action in state,
```
typo, it should be "Return the set of ..."

simfis replied:
Fixed.

mensch72 commented:
Hopefully the merge is not too difficult...

mensch72 commented:
parameters cannot be changed during the lifetime of the agent at the moment.

Am 9. April 2024 20:17:52 MESZ schrieb Massimo Redaelli ***@***.***>:
>@mredaelli commented on this pull request.
>
>
>
>> +        return res
>+
>+    @cache
>+    def randomTieBreaker(self, state, action):
>+        return random.random()
>+
>+    # now we can combine all of the above quantities to a combined (safety) loss function:
>+
>+    # state, action, aleph4state, aleph4action, estActionProbability
>+    @lru_cache(maxsize=None)
>+    def combinedLoss(self, s, a, al4s, al4a, p): # recursive
>+
>+        if self.debug:
>+            print(pad(s),"| | combinedLoss, state",prettyState(s),"action",a,"aleph4state",al4s,"aleph4action",al4a,"estActionProbability",p,"...")
>+
>+        res = sum(k * v(s, a, al4s, al4a, p) for k, v in self.losses.items())
>
>If I'm not mistaken, that's done in `__init__`, here:
>
>https://github.com/pik-gane/satisfia/pull/7/files#diff-2a77c67961414a5d12c1a7db8c3b4536e945a06643c4f16f68f3509b002378eeR190
>
>so we do it only once rather than at every evaluation.
>
>Unless the parameters can be changed?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/pik-gane/satisfia/pull/7#discussion_r1558109685
>You are receiving this because your review was requested.
>
>Message ID: ***@***.***>

mredaelli commented:
rebased and removed some numpy.

Down to:
```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):     22.559 s ±  0.579 s    [User: 21.415 s, System: 1.334 s]
  Range (min … max):   21.661 s … 23.575 s    10 runs
```

mredaelli commented:
Added a commit to remove `prev_loc` from state.

```
❯ hyperfine "python scripts/perf_eval.py"
Benchmark 1: python scripts/perf_eval.py
  Time (mean ± σ):      5.990 s ±  0.359 s    [User: 5.750 s, System: 0.521 s]
  Range (min … max):    5.670 s …  6.830 s    10 runs
```
